core/src/androidMain/kotlin/Peripheral.kt

```diff
@@ -78,34 +77,39 @@ public class AndroidPeripheral internal constructor(
         combine(_ready, state) { ready, state -> ready && state == State.Connected }.first { it }
     }
 
+    private fun establishConnection(): Connection =
+        bluetoothDevice.connect(
+            context,
+            _state,
+            invokeOnClose = { connectJob.value = null }
+        ) ?: throw ConnectionRejectedException()
+
+    /** Creates a connect [Job] that completes when connection is established, or failure occurs. */
     private fun createConnectJob(): Job = scope.launch(start = LAZY) {
         _ready.value = false
 
-        val connection =
-            bluetoothDevice.connect(context, _state) ?: throw ConnectionRejectedException()
-        _connection = connection
+        val connection = establishConnection().also { _connection = it }
+        connection
+            .characteristicChanges
+            .onEach(observers.characteristicChanges::emit)
+            .launchIn(scope, start = UNDISPATCHED)
```

Comment:
Are you intentionally not catching the errors that could be thrown from this?

Reply:
Correct.

Before this PR, upstream the `Channel` was being closed with an exception, only to have it catch here as a signal to stop collecting.
Instead, just performing a standard close (without exception) upstream so that this stops collection gracefully.